rest/main.go

```diff
@@ -102,10 +108,31 @@ func serverMainPersistentConfig(fs *flag.FlagSet, flagStartupConfig *StartupConf
 	if len(configPath) == 1 {
 		fileStartupConfig, err := LoadStartupConfigFromPath(configPath[0])
 		if pkgerrors.Cause(err) == base.ErrUnknownField {
-			// TODO: CBG-1399 Do automatic legacy config upgrade here
-			base.Warnf("Couldn't parse bootstrap config and legacy config migration not yet implemented: %v", err)
-			// When automatic legacy config upgrade is done return when disable_persistent_config=true
-			return true, nil
+			// If we have an unknown field error processing config its possible that the config is a 2.x config
+			// requiring automatic upgrade. We should attempt to perform this upgrade
+
+			var upgradeError error
+			fileStartupConfig, disablePersistentConfigFallback, upgradeError = automaticConfigUpgrade(configPath[0])
+			if upgradeError != nil {
+
+				// We need to validate if the error was again, an unknown field error. If this is the case its possible
+				// the config is actually a 3.x config but with a genuine unknown field, therefore we should  return the
+				// original error from LoadStartupConfigFromPath.
+				if pkgerrors.Cause(upgradeError) == base.ErrUnknownField {
+					return false, err
+				}
+
+				return false, upgradeError
+			}
+
+			if disablePersistentConfigFallback {
+				return true, nil
+			}
+
+			// If we have got this far the automatic config upgrade was successful so we should nil out the error so we
+			// can continue
+			err = nil
+
 		}
 		if err != nil {
```

Comment:
This is effectively the same thing now we're not always returning, feels cleaner and more obvious

```suggestion
		} else if err != nil {
```